### ISSUE TITLE
Combine multiple matching algorithms and patterns with booleans

### ISSUE THREAD

**Issue by alexzeitgeist** [jonaswinkler/paperless-ng#447]
It would be nice if we could use more than one matching algorithm per correspondents/tag/type. These matching algorithms could then be combined with booleans such as AND/NOT/OR.

**Comment by jonaswinkler** [comment]
Not sure if this is really required by many users.

If, at all, I could imagine adding that as a new matching algorithm, lets say "Complex match", and build a small DSL for that. Need to look for frameworks in python. I've done things like that many times in other languages.

Matching Algorithm: `Complex match`
Match: `( any("a b c") OR all("d e f") ) AND NOT regex("[0-9][abc]")`

Definitely something intented for power users only. Building a new UI for that with visual and/or/not selection of multiple matching algorithms is a rather complex change compared to that, and requires database changes and migrations.

**Comment by alexzeitgeist** [comment]
I agree for the reasons you cited that it'd be best to add this as a new matching algorithm.

Personally I encounter the situation relatively often where, for example, I would like to make an exception (AND NOT) to a matching rule to narrow down the match and avoid false positives. I could try with regex, but that's more tedious and prone to errors.

**Comment by jonaswinkler** [comment]
I see the issues with that; That's also part of the reason I added machine learning matching a couple years back. I found it pretty hard to define good matches with the given algorithms that could handle both
* unrelated documents that did contain the match
* related documents that did not contain the match (due to OCR errors, for example)

The auto algorithm is able to detect and use many terms both inclusive and exclusive, but I also know that there are cases where you'd want more control over the matching.

**Comment by ignoreigor** [comment]
Hi,
I would also have liked the possibility of a more complex set of rules. But that's only a feeling. No idea, how much the benefit will finally be.
What about matching an expression in dependency on a tag or a document-type or a correspondent? 

But I think, there is a "Henne-Ei-Problem", because this metadata is also set by a rule. So you have to define priorities, this is perhaps again too much effort for the benefit it would have for some users.

I think I can work with regular expressions or will give the automatic algorithm a chance.

**Comment by jonaswinkler** [comment]
> What about matching an expression in dependency on a tag or a document-type or a correspondent?

This would essentially be the same as combining two different matching algorithms with an AND construct.

**Comment by alexzeitgeist** [comment]
I admit I haven't used the auto algorithm yet, mainly because most of my documents are rather easy to identify by the other algorithms.

Another reason why multiple matching algorithms might make sense: I have various documents that are essentially different, but that should have one of more identical tags attached to them. This isn't possible right now, because only one matching algorithm rule can be used for the tag, and the same tag can be defined only once.

**Comment by ignoreigor** [comment]
Hi,

> > What about matching an expression in dependency on a tag or a document-type or a correspondent?
> 
> This would essentially be the same as combining two different matching algorithms with an AND construct.

yes you're right. Of course it is an AND-Condition at last.
But it might be easier and you can do multiple actions with an easy rule instead of joining to conditions. 

@alexzeitgeist 's idea could solve an OR. Define multiple rules for a tag.
Problem is to get a working mechanism to compute the rules in an order.

Imagine: Do a tagging by matching a rule only when a document was classified as a specific document-type from a single correspondent before.

**Comment by sebastientaggart** [comment]
One example I have that the existing matching can't do (or I don't know how without regex) is:

I want to match against an address, and add a tag if found.  So I'd like to tag `Home` if the address `123 Oak St` is found.  However, the address sometimes comes as `123 Oak St` OR `123 OAK ST` OR `123 OAK STREET`, etc.  Is there a way to do this?

**Comment by jonaswinkler** [comment]
If you use All, and specify "123 OAK STREET", it will only match documents containing all these words, in any order, or separated by other words. Probably gets you closest to what you want to do.

**Comment by blackfurball** [comment]
I too would like to see the possibility to have some kind of negative matching in a pattern.

For example: I use Tags to categorize invoices nad correspondance for insurance, car repairs, health spendings, IT-Stuff and so on.

The problem is: As soon as I upload my banks Kontoauszug ("statement of account"), paperlessng will add pretty much every tag to it I only want to tag invoices and correspondance with. 

The possibility for something like: Any Word in ("InsuranceCorp inc." Insurance" "CarInsurance" -Kontoauszug) would be amazing for this problem.
(Notice the "-" in front of Kontoauszug as a negative match)

Otherwise I'm fairly satisfied with what paperlessng does for me :)
